framework/src/transfers/MultiAppReporterTransfer.C

```diff
@@ -0,0 +1,113 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MultiAppReporterTransfer.h"
+
+registerMooseObject("MooseApp", MultiAppIntegerReporterTransfer);
+registerMooseObject("MooseApp", MultiAppRealReporterTransfer);
+registerMooseObject("MooseApp", MultiAppVectorReporterTransfer);
+registerMooseObject("MooseApp", MultiAppStringReporterTransfer);
+
+template <typename ReporterType>
+InputParameters
+MultiAppReporterTransfer<ReporterType>::validParams()
+{
+  InputParameters params = MultiAppTransfer::validParams();
+  params.addClassDescription(
+      "Transfers reporter data between the master application and sub-application(s).");
+  params.addRequiredParam<std::vector<ReporterName>>(
+      "from_reporters",
+      "List of the reporter names (object_name/value_name) to transfer the value from.");
+  params.addRequiredParam<std::vector<ReporterName>>(
+      "to_reporters",
+      "List of the reporter names (object_name/value_name) to transfer the value to.");
+  params.addParam<unsigned int>(
+      "subapp_index",
+      std::numeric_limits<unsigned int>::max(),
+      "The MultiApp object sub-application index to use when transferring to/from the "
+      "sub-application. If unset and transferring to the sub-applications then all "
+      "sub-applications will receive data. The value must be set when transferring from a "
+      "sub-application.");
+  return params;
+}
+
+template <typename ReporterType>
+MultiAppReporterTransfer<ReporterType>::MultiAppReporterTransfer(const InputParameters & parameters)
+  : MultiAppTransfer(parameters),
+    _from_reporter_names(getParam<std::vector<ReporterName>>("from_reporters")),
+    _to_reporter_names(getParam<std::vector<ReporterName>>("to_reporters")),
+    _subapp_index(getParam<unsigned int>("subapp_index"))
+{
+  if (_from_reporter_names.size() != _to_reporter_names.size())
+    paramError("to_reporters", "from_reporters and to_reporters must be the same size.");
+}
+
+template <typename ReporterType>
+void
+MultiAppReporterTransfer<ReporterType>::executeToMultiapp()
+{
+  for (unsigned int n = 0; n < _from_reporter_names.size(); ++n)
+  {
+    const ReporterType & value =
+        _multi_app->problemBase().template getReporterValueByName<ReporterType>(
+            _from_reporter_names[n]);
+
+    if (_subapp_index == std::numeric_limits<unsigned int>::max())
+    {
+      for (unsigned int i = 0; i < _multi_app->numGlobalApps(); ++i)
+        if (_multi_app->hasLocalApp(i))
+          _multi_app->appProblemBase(i).template setReporterValueByName<ReporterType>(
+              _to_reporter_names[n], value);
+    }
+
+    else if (_multi_app->hasLocalApp(_subapp_index))
+      _multi_app->appProblemBase(_subapp_index)
+          .template setReporterValueByName<ReporterType>(_to_reporter_names[n], value);
```

Comment:
Based on my comment above, just add MultiAppReporterTransfer to the friends of the key object in ReporterData and have his use the getReporterData method that takes that key object.